File: depot/src/main/kotlin/org/goodmath/polytope/depot/stashes/ChangeStash.kt
/*
 * Copyright 2023 Mark C. Chu-Carroll
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.goodmath.polytope.depot.stashes

import maryk.rocksdb.ColumnFamilyHandle
import maryk.rocksdb.RocksDB
import org.goodmath.polytope.Config
import org.goodmath.polytope.common.*
import org.goodmath.polytope.common.stashable.*
import org.goodmath.polytope.depot.Depot
import org.goodmath.polytope.depot.util.*
import java.time.Instant


class ChangeStash(
    private val db: RocksDB,
    private val changesColumn: ColumnFamilyHandle,
    private val savePointsColumn: ColumnFamilyHandle,
    private val depot: Depot
): Stash {

    /**
     * Retrieve a change.
     * @param auth the user performing the operation.
     * @param project the project containing the change.
     * @param id the change ID.
     * @return the change
     */
    private fun retrieveChange(auth: AuthenticatedUser, project: String, id: Id<Change>): Change {
        depot.users.validatePermissions(auth, Action.readProject(project))
        val c = db.getTyped<Change>(changesColumn, id) ?: throw PtException(
            PtException.Kind.NotFound,
            "Change $id not found"
        )
        if (c.project != project) {
            throw PtException(
                PtException.Kind.InvalidParameter,
                "Change $id doesn't belong to project $project"
            )
        }
        return c
    }

    /**
     * Get the ID for a named change.
     * @param project the project containing the change.
     * @param history the name of the history containing the change.
     * @param changeName the name of the change.
     * @return the change ID.
     * @throws PtException if no change with the name is found.
     */
    private fun getChangeId(project: String, history: String, changeName: String): Id<Change> {
        val idx = getCollectionIndex()
        return idx.entries[ChangeIndex.ChangeIndexKey(project, history, changeName)] ?: throw PtException(
            PtException.Kind.NotFound,
            "Change named $changeName not found in project $project"
        )
    }

    /**
     * Retrieve a change by name.
     * @param auth the user performing the operation.
     * @param project the project containing the change.
     * @param history the name of the history containing the change
     * @param changeName the name of the change.
     * @returns the change
     * @throws PtException if no change with the name is found.
     */
    fun retrieveChangeByName(auth: AuthenticatedUser, project: String, history: String, changeName: String): Change {
        depot.users.validatePermissions(auth, Action.writeProject(project))
        val changeId = getChangeId(project, history, changeName)
        return retrieveChange(auth, project, changeId)
    }

    /**
     * Create a new change
     * @param auth the user performing the operation.
     * @param projectName the project containing the change.
     * @param history the name of the history that will contain the change.
     * @param changeName the name of the new change.
     * @param basis the project version specifier of the baseline of the change.
     * @param description a description of the change.
     * @return the new change.
     */
    fun createChange(
        auth: AuthenticatedUser,
        projectName: String,
        history: String,
        changeName: String,
        basis: ProjectVersionSpecifier,
        description: String
    ): Change {
        depot.users.validatePermissions(auth, Action.writeProject(projectName))
        val project = depot.projects.retrieveProject(auth, projectName)
        val change = Change(
            id = newId<Change>(ID_CHANGE),
            project = projectName,
            name = changeName,
            history = history,
            baseline = project.baseline,
            basis = basis,
            description = description,
            savePoints = emptyList(),
            status = ChangeStatus.Open,
            timestamp = Instant.now().toEpochMilli()
        )
        db.putTyped(changesColumn, change.id, change)
        updateCollectionIndex(projectName, history, changeName, change.id)
        return change
    }

    /**
     * Update the status of an in-progress change.
     * @param auth the user performing the operation.
     * @param project the project containing the change.
     * @param history the history containing the change.
     * @param changeName the name of the change.
     * @param status the new status of the change.
     */
    fun updateChangeStatus(
        auth: AuthenticatedUser,
        project: String,
        history: String,
        changeName: String,
        status: ChangeStatus
    ) {
        depot.users.validatePermissions(auth, Action.writeProject(project))
        val changeId = getChangeId(project, history, changeName)
        val change = retrieveChange(auth, project, changeId)
        val updated = change.copy(timestamp = Instant.now().toEpochMilli(),
            status = status)
        db.updateTyped(changesColumn, changeId, updated)
    }

    /**
     * Get the status of a change.
     * @param auth the user performing the operation.
     * @param project the project containing the change.
     * @param history the history containing the change.
     * @param changeName the name of the change.
     * @return the change status of the change.
     */
    fun getChangeStatus(
        auth: AuthenticatedUser,
        project: String, history: String, changeName: String
    ): ChangeStatus {
        depot.users.validatePermissions(auth, Action.readProject(project))
        val change = retrieveChangeByName(auth, project, history, changeName)
        return change.status
    }

    /**
     * Create a save point in an open change.
     * @param auth the user performing the operation.
     * @param project the project containing the change.
     * @param history the name of the history containing the change.
     * @param changeName the name of the change.
     * @param changedArtifacts a list of the artifacts that were modified since the last
     *    baseline.
     * @param description a description of the savepoint.
     * @param basis a PVS for the starting point of the change.
     * @param baselineVersion the version ID of the new baseline.
     * @return a new savepoint.
     */
    fun createSavePoint(
        auth: AuthenticatedUser,
        project: String,
        history: String,
        changeName: String,
        changedArtifacts: List<Id<Artifact>>,
        description: String,
        basis: ProjectVersionSpecifier,
        baselineVersion: Id<ArtifactVersion>
    ): SavePoint {
        depot.users.validatePermissions(auth, Action.writeProject(project))
        val ch = retrieveChangeByName(auth, project, history, changeName)
        val savePoint = SavePoint(
            id = newId<SavePoint>("cs"),
            changeId = ch.id,
            modifiedArtifacts = changedArtifacts,
            basis = basis,
            baselineVersion = baselineVersion,
            creator = auth.userId,
            description = description,
            timestamp = Instant.now().toEpochMilli()
        )
        val updatedChange = ch.copy(savePoints = ch.savePoints + savePoint.id)
        db.putTyped(savePointsColumn, savePoint.id, savePoint)
        db.updateTyped(changesColumn, updatedChange.id, updatedChange)
        return savePoint
    }

    /**
     * Retrieve a savepoint
     * @param auth the user performing the operation.
     * @param project the project containing the change.
     * @param history the name of the history containing the change.
     * @param changeName the name of the change.
     * @param
     *
     */
    fun retrieveSavePoint(
        auth: AuthenticatedUser,
        project: String,
        history: String,
        changeName: String,
        saveId: Id<SavePoint>
    ): SavePoint {
        depot.users.validatePermissions(auth, Action.readProject(project))
        val ch = retrieveChangeByName(auth, project, history, changeName)
        return if (ch.savePoints.contains(saveId)) {
            db.getTyped<SavePoint>(savePointsColumn, saveId)
                ?: throw PtException(
                    PtException.Kind.NotFound,
                    "Save point not found"
                )
        } else {
            throw PtException(
                PtException.Kind.NotFound,
                "Save point not found"
            )
        }
    }

    /**
     * List the save points in a change.
     * @param auth the user performing the operation.
     * @param project the project containing the change.
     * @param history the name of the history containing the change.
     * @param change the name of the change.
     * @return a list of save point IDs.
     */
    fun listSavePoints(
        auth: AuthenticatedUser,
        project: String,
        history: String,
        change: String
    ): List<SavePoint> {
        depot.users.validatePermissions(auth, Action.readProject(project))
        val ch = retrieveChangeByName(auth, project, history, change)
        return ch.savePoints.map { retrieveSavePoint(auth, project, history, change, it) }
    }

    /**
     * List the changes in a history.
     * @param auth the user performing the operation.
     * @param project the project containing the change.
     * @param historyName the name of the history.
     * @return a list of changes.
     */
    fun listChanges(
        auth: AuthenticatedUser,
        project: String,
        historyName: String
    ): List<Change> {
        depot.users.validatePermissions(auth, Action.readProject(project))
        val index = getCollectionIndex()
        return index.entries.entries.filter { it.key.project == project && it.key.history == historyName }
            .map { entry -> retrieveChange(auth, project, entry.value)}

    }

    override fun initStorage(config: Config) {
        val cIdx = ChangeIndex(mutableMapOf())
        db.putTyped(changesColumn, indexKey, cIdx)
    }

    /**
     * An index of changes, which lets us find and list changes by name without having
     * to scan all changes in the DB.
     */
    data class ChangeIndex(
        val entries: MutableMap<ChangeIndexKey, Id<Change>>
    ) {
        data class ChangeIndexKey(val project: String, val history: String, val changeName: String)

        fun get(projectName: String, historyName: String, changeName: String): Id<History>? {
            return entries[ChangeIndexKey(projectName, historyName, changeName)]
        }

        fun add(projectName: String, historyName: String, changeName: String, id: Id<Change>) {
            entries[ChangeIndexKey(projectName, historyName, changeName)] = id
        }
    }

    private val indexKey = "__change__index__"
    private var changeIndex: ChangeIndex? = null
    private fun getCollectionIndex(): ChangeIndex {
        if (changeIndex == null) {
            changeIndex = db.getTyped(changesColumn, indexKey)
        }
        return changeIndex!!
    }

    private fun updateCollectionIndex(project: String, history: String, changeName: String, id: Id<History>) {
        val idx = getCollectionIndex()
        idx.add(project, history, changeName, id)
        db.updateTyped(changesColumn, indexKey, idx)
    }



}

File: src/polytope/depot/stashes/change_stash.py
# Copyright 2025 Mark C. Chu-Carroll
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


from datetime import datetime
from typing import List
from polytope.common.error import ErrorKind, PtException
from polytope.common.stashable.artifact import Artifact, ArtifactVersion
from polytope.common.stashable.change import Change, ChangeStatus, SavePoint
from polytope.common.stashable.ids import Id, IdKind
from polytope.common.stashable.pvs import ProjectVersionSpecifier
from polytope.common.stashable.user import Action, AuthenticatedUser
from polytope.depot.depot import Depot
from polytope.depot.stashes.stash import Stash
from polytope.depot.stashes.user_stash import UserStash


class ChangeStash(Stash):
    """
    The stash for storing changes and savepoints within changes.

    Storage schema:

        stash[CHANGE][project: str][history: str][id: Id[Change]]: Change
        stash[CHANGE_INDEX][project: str][history: str][name: str]: Id[Change]
        stash[SAVEPOINT][project: str][history: str][id: Id[SavePoint]]: SavePoint
    """

    def __init__(self, depot: Depot) -> None:
        self.depot = depot
        self.changes = depot.db.get_collection("changes")
        self.saves = depot.db.get_collection("saves")

    def init_storage(self, config):
        pass

    @property
    def user_stash(self) -> UserStash:
        return self.depot.user_stash

    def retrieve_change(
            self,
            auth: AuthenticatedUser,
            project: str,
            id: Id[Change]) -> Change:
        """
        Retrieve a change.
        auth -- the user performing the operation.
        project -- the project containing the change.
        id -- the change ID.
        @return the change
        """
        self.user_stash.validate_permissions(auth, Action.read_project(project))
        ch_dict = self.changes.find_one({"_id": id,
                                         "project": project
                                         })
        if ch_dict is None:
            raise PtException(ErrorKind.NotFound,
                              f"Change {id} not found")
        return Change.from_dict(ch_dict)

    def get_change_id(
            self,
            auth: AuthenticatedUser,
            project: str,
            history: str,
            change_name: str) -> Id[Change]:
        """
        Get the ID for a named change.

        Arguments:
        project -- the project containing the change.
        history -- the name of the history containing the change.
        change_name -- the name of the change.

        Returns the change ID.
        """
        ch = self.retrieve_change_by_name(auth, project, history, change_name)
        return ch.id

    def retrieve_change_by_name(
            self,
            auth: AuthenticatedUser,
            project: str,
            history: str,
            change_name: str) -> Change:
        """
        Retrieve a change by name.

        Arguments:
        auth -- the user performing the operation.
        project -- the project containing the change.
        history -- the name of the history containing the change
        change_name -- the name of the change.

        Returns the change
        """
        self.user_stash.validate_permissions(auth, Action.read_project(project))
        ch_dict = self.changes.find_one({"project": project,
                                         "history": history,
                                         "name": change_name})
        if ch_dict is None:
            raise PtException(ErrorKind.NotFound,
                              f"Change named {change_name} not found in history {history} of project {project}")
        return Change.from_dict(ch_dict)

    def create_change(
        self,
        auth: AuthenticatedUser,
        project_name: str,
        history: str,
        change_name: str,
        basis: ProjectVersionSpecifier,
        description: str
    ) -> Change:
        """
        Create a new change

        Arguments:
        auth -- the user performing the operation.
        project_name -- the project containing the change.
        history -- the name of the history that will contain the change.
        change_name -- the name of the new change.
        basis -- the project version specifier of the baseline of the change.
        description -- a description of the change.

        Returns the new change.
        """

        self.user_stash.validate_permissions(auth, Action.write_project(project_name))
        project = self.depot.project_stash.retrieve_project(auth, project_name)
        baseline = self.depot.project_stash.resolve_project_version_specifier(auth, basis)

        change = Change(
            id=Id.new_id(IdKind.ID_CHANGE),
            project=project_name,
            name=change_name,
            history=history,
            baseline=baseline,
            basis=basis,
            description=description,
            save_points=[],
            status=ChangeStatus.Open,
            timestamp=datetime.now()
        )
        self.changes.insert_one(change.to_dict())
        return change

    def update_change_status(
        self,
        auth: AuthenticatedUser,
        project: str,
        history: str,
        change_name: str,
        status: ChangeStatus
    ) -> None:
        """
        Update the status of an in-progress change.
        auth -- the user performing the operation.
        project -- the project containing the change.
        history -- the history containing the change.
        change_name -- the name of the change.
        status -- the new status of the change.
        """
        self.user_stash.validate_permissions(auth, Action.write_project(project))
        update = self.changes.update_one({"project": project,
                                          "history": history,
                                          "name": change_name,
                                          },
                                         {"$set": {
                                             "status": status.value
                                         }})
        if update.modified_count != 1:
            raise PtException(ErrorKind.NotFound,
                              f"Change named {change_name} not found in history {history} of project {project}")

    def create_save_point(
        self,
        auth: AuthenticatedUser,
        project: str,
        history: str,
        change_name: str,
        changed_artifacts: List[Id[Artifact]],
        description: str,
        basis: ProjectVersionSpecifier,
        baseline_version: Id[ArtifactVersion]
    ) -> SavePoint:
        """
        Create a save point in an open change.

        Arguments:
        auth - - the user performing the operation.
        project - - the project containing the change.
        history - - the name of the history containing the change.
        change_name - - the name of the change.
        changed_artifacts - - a list of the artifacts that were modified since the last
            baseline.
        description - - a description of the savepoint.
        basis - - a PVS for the starting point of the change.
        baseline_version - - the version ID of the new baseline.

        Returns a new savepoint.
        """
        self.user_stash.validate_permissions(auth, Action.write_project(project))
        ch = self.retrieve_change_by_name(auth, project, history, change_name)
        save_point = SavePoint(
            id=Id.new_id(IdKind.ID_CHANGE_SAVE),
            change_id=ch.id,
            modified_artifacts=changed_artifacts,
            basis=basis,
            baseline_version=baseline_version,
            creator=auth.user_id,
            description=description,
            timestamp=datetime.now()
        )
        self.saves.insert_one(save_point.to_dict())
        self.changes.update_one({"_id": str(ch.id)},
                                {"$push": {
                                    "save_points": str(save_point.id)
                                }})
        return save_point

    def retrieve_save_point(
        self,
        auth: AuthenticatedUser,
        project: str,
        history: str,
        change_name: str,
        save_id: Id[SavePoint]
    ) -> SavePoint:
        """
        Retrieve a savepoint

        Arguments:
        auth -- the user performing the operation.
        project -- the project containing the change.
        history -- the name of the history containing the change.
        change_name -- the name of the change.
        save_id -- the ID of the savepoint.

        Returns the savepoint
        """
        self.user_stash.validate_permissions(auth, Action.read_project(project))
        change_id = self.get_change_id(auth, project, history, change_name)
        sp_dict = self.saves.find_one({"_id": str(save_id), "change_id": change_id})
        if sp_dict is None:
            raise PtException(ErrorKind.NotFound, f"Save point not found in change {change_name}")
        return SavePoint.from_dict(sp_dict)

    def list_save_points(
        self,
        auth: AuthenticatedUser,
        project: str,
        history: str,
        change: str
    ) -> List[Id[SavePoint]]:
        """
        List the save points in a change.

        Arguments:
        auth - - the user performing the operation.
        project - - the project containing the change.
        history - - the name of the history containing the change.
        change - - the name of the change.

        Returns a list of save point IDs.
        """
        self.user_stash.validate_permissions(auth, Action.read_project(project))
        ch = self.retrieve_change_by_name(auth, project, history, change)
        return ch.save_points

    def list_changes(
        self,
        auth: AuthenticatedUser,
        project: str,
        history_name: str,
        show: ChangeStatus
    ) -> List[Change]:
        """
        List the changes in a history.

        Arguments:
        auth -- the user performing the operation.
        project -- the project containing the change.
        history_name -- the name of the history.
        show -- the minimum status to include in the result list. If this is "Aborted", then
           all changes, open, closed, and aborted will be included; if it's "Closed", then both open
           and closed will be included, but aborted changes will not; if it's "Open", then all changes
           will be included

        Returns a list of changes.
        """
        self.user_stash.validate_permissions(auth, Action.read_project(project))
        changes = self.changes.find({"project": project, "history": history_name})
        result: List[Change] = []
        for ch in changes:
            match show:
                case ChangeStatus.Aborted:
                    result.append(Change.from_dict(ch))
                case ChangeStatus.Closed:
                    if ch.status == ChangeStatus.Closed or ch.status == ChangeStatus.Open:
                        result.append(Change.from_dict(ch))
                case ChangeStatus.Open:
                    if ch.status == ChangeStatus.Open:
                        result.append(Change.from_dict(ch.to_change()))
        return result
